File: Scripts/performance_analyzer.py
#!/usr/bin/env python3
"""
Performance Analyzer for Swift Projects
A modular, reusable tool for analyzing performance patterns in Swift codebases
"""

import os
import re
import json
import argparse
from datetime import datetime
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass, asdict
from enum import Enum
from pathlib import Path


class Severity(Enum):
    """Issue severity levels"""
    CRITICAL = "Critical"
    HIGH = "High"
    MEDIUM = "Medium"
    LOW = "Low"
    INFO = "Info"


class Category(Enum):
    """Performance issue categories"""
    MEMORY = "Memory Management"
    CONCURRENCY = "Concurrency"
    ALGORITHM = "Algorithm Complexity"
    IO = "I/O Operations"
    UI = "UI Responsiveness"
    NETWORK = "Network"
    DATABASE = "Database"


class Impact(Enum):
    """Performance impact levels"""
    HIGH = "High Performance Impact"
    MEDIUM = "Medium Performance Impact"
    LOW = "Low Performance Impact"
    NEGLIGIBLE = "Negligible"


@dataclass
class PerformanceMetric:
    """Represents a single performance issue or metric"""
    name: str
    category: Category
    severity: Severity
    file: str
    line: Optional[int]
    description: str
    recommendation: Optional[str]
    estimated_impact: Impact
    
    def to_dict(self):
        """Convert to dictionary for JSON serialization"""
        return {
            'name': self.name,
            'category': self.category.value,
            'severity': self.severity.value,
            'file': self.file,
            'line': self.line,
            'description': self.description,
            'recommendation': self.recommendation,
            'estimated_impact': self.estimated_impact.value
        }


class PerformanceAnalyzer:
    """Main analyzer class for Swift performance patterns"""
    
    # Pattern definitions for various performance issues
    PATTERNS = {
        'retain_cycles': [
            (r'\{\s*\[self\]', "Strong self capture in closure"),
            (r'\{[\s]*\[self\]', "Strong self capture in closure"),
            (r'Timer\.scheduledTimer.*target:\s*self', "Timer with strong self reference"),
            (r'NotificationCenter\.default\.addObserver\(self(?!.*removeObserver)', "Notification observer without removal")
        ],
        'memory': [
            (r'Data\(contentsOf:(?!.*async)', "Synchronous data loading"),
            (r'UIImage\(data:(?!.*async)', "Synchronous image loading"),
            (r'class\s+\w+.*\{(?!.*deinit)', "Class without deinit"),
            (r'\.copy\(\)(?!.*autoreleasepool)', "Copy without autorelease pool")
        ],
        'concurrency': [
            (r'DispatchQueue\.main\.sync', "Main thread blocking"),
            (r'Task\.detached', "Unstructured concurrency"),
            (r'@Published(?!.*@MainActor)', "Published without MainActor"),
            (r'DispatchSemaphore', "Semaphore usage (consider async/await)"),
            (r'\.wait\(\)', "Blocking wait operation")
        ],
        'algorithm': [
            (r'for\s+.*\s+in\s+.*\{[^}]*for\s+.*\s+in', "Nested loops detected"),
            (r'\.sorted\(\)\.first', "Sorting for single element"),
            (r'\.filter\(.*\)\.count\s*==\s*0', "Inefficient empty check"),
            (r'\.map\(.*\)\.filter\(.*\)\.reduce', "Multiple collection operations"),
            (r'Array\(repeating:.*count:\s*\d{4,}', "Large array allocation")
        ],
        'io': [
            (r'try\s+String\(contentsOf:(?!.*async)', "Synchronous file reading"),
            (r'try\s+Data\(contentsOf:(?!.*async)', "Synchronous data loading"),
            (r'UserDefaults\.standard\.\w+', "UserDefaults access"),
            (r'FileManager\.default\.(?:createFile|removeItem|copyItem)(?!.*async)', "Synchronous file operations")
        ],
        'ui': [
            (r'\.onAppear\s*\{[^}]*\.(sorted|filter|map|reduce)', "Heavy operation in onAppear"),
            (r'body\s*:\s*some\s+View\s*\{[^}]*for\s+.*\s+in', "Loop in SwiftUI body"),
            (r'\.task\s*\{(?!.*await)', "Task without await"),
            (r'Image\(uiImage:(?!.*async)', "Synchronous image creation")
        ]
    }
    
    def __init__(self, project_path: str):
        """Initialize analyzer with project path"""
        self.project_path = Path(project_path)
        self.metrics: List[PerformanceMetric] = []
        
    def analyze(self) -> Dict:
        """Run full analysis on the project"""
        print(f"🔍 Starting performance analysis for: {self.project_path}")
        
        # Find all Swift files
        swift_files = self._find_swift_files()
        print(f"📁 Found {len(swift_files)} Swift files to analyze")
        
        # Analyze each file
        for file_path in swift_files:
            self._analyze_file(file_path)
        
        # Generate summary
        summary = self._generate_summary()
        
        return {
            'timestamp': datetime.now().isoformat(),
            'project_path': str(self.project_path),
            'metrics': [m.to_dict() for m in self.metrics],
            'summary': summary
        }
    
    def _find_swift_files(self) -> List[Path]:
        """Find all Swift files in the project"""
        swift_files = []
        
        for root, dirs, files in os.walk(self.project_path):
            # Skip build and test directories
            dirs[:] = [d for d in dirs if d not in ['.build', 'DerivedData', '.git']]
            
            for file in files:
                if file.endswith('.swift') and 'Tests' not in root:
                    swift_files.append(Path(root) / file)
        
        return swift_files
    
    def _analyze_file(self, file_path: Path):
        """Analyze a single Swift file"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
                lines = content.split('\n')
        except Exception as e:
            print(f"⚠️  Error reading {file_path}: {e}")
            return
        
        file_name = file_path.name
        
        # Check for retain cycles
        self._check_patterns(content, lines, file_name, 'retain_cycles', 
                           Category.MEMORY, Severity.HIGH, Impact.HIGH,
                           "Use [weak self] or [unowned self] to break retain cycles")
        
        # Check memory management
        self._check_patterns(content, lines, file_name, 'memory',
                           Category.MEMORY, Severity.MEDIUM, Impact.MEDIUM,
                           "Consider async operations or proper memory management")
        
        # Check concurrency issues
        self._check_patterns(content, lines, file_name, 'concurrency',
                           Category.CONCURRENCY, Severity.HIGH, Impact.HIGH,
                           "Use modern Swift concurrency features")
        
        # Check algorithm complexity
        self._check_patterns(content, lines, file_name, 'algorithm',
                           Category.ALGORITHM, Severity.MEDIUM, Impact.MEDIUM,
                           "Optimize algorithm for better performance")
        
        # Check I/O operations
        self._check_patterns(content, lines, file_name, 'io',
                           Category.IO, Severity.MEDIUM, Impact.MEDIUM,
                           "Use async I/O operations")
        
        # Check UI responsiveness
        if 'View' in file_name or 'ViewModel' in file_name:
            self._check_patterns(content, lines, file_name, 'ui',
                               Category.UI, Severity.HIGH, Impact.HIGH,
                               "Move heavy operations out of UI code")
        
        # Additional checks
        self._check_specific_issues(content, lines, file_name)
    
    def _check_patterns(self, content: str, lines: List[str], file_name: str,
                       pattern_key: str, category: Category, severity: Severity,
                       impact: Impact, recommendation: str):
        """Check for specific patterns in the content"""
        for pattern, description in self.PATTERNS[pattern_key]:
            matches = re.finditer(pattern, content, re.MULTILINE | re.DOTALL)
            for match in matches:
                line_num = self._get_line_number(match.start(), content)
                
                self.metrics.append(PerformanceMetric(
                    name=f"{pattern_key.replace('_', ' ').title()} Issue",
                    category=category,
                    severity=severity,
                    file=file_name,
                    line=line_num,
                    description=description,
                    recommendation=recommendation,
                    estimated_impact=impact
                ))
    
    def _check_specific_issues(self, content: str, lines: List[str], file_name: str):
        """Check for specific performance issues"""
        
        # Check for excessive UserDefaults usage
        userdefaults_count = len(re.findall(r'UserDefaults\.standard', content))
        if userdefaults_count > 5:
            self.metrics.append(PerformanceMetric(
                name="Excessive UserDefaults Access",
                category=Category.IO,
                severity=Severity.LOW,
                file=file_name,
                line=None,
                description=f"File has {userdefaults_count} UserDefaults accesses",
                recommendation="Cache UserDefaults values in properties",
                estimated_impact=Impact.LOW
            ))
        
        # Check for large class/struct
        if len(lines) > 500:
            self.metrics.append(PerformanceMetric(
                name="Large File",
                category=Category.ALGORITHM,
                severity=Severity.LOW,
                file=file_name,
                line=None,
                description=f"File has {len(lines)} lines, consider splitting",
                recommendation="Break down into smaller, focused components",
                estimated_impact=Impact.LOW
            ))
        
        # Check for missing async in network calls
        if 'URLSession' in content and 'async' not in content:
            self.metrics.append(PerformanceMetric(
                name="Synchronous Network Call",
                category=Category.NETWORK,
                severity=Severity.CRITICAL,
                file=file_name,
                line=None,
                description="URLSession usage without async/await",
                recommendation="Use async/await for network operations",
                estimated_impact=Impact.HIGH
            ))
        
        # Check for force unwrapping in production code
        force_unwrap_count = len(re.findall(r'!\s*[,\)\}\.]', content))
        if force_unwrap_count > 3:
            self.metrics.append(PerformanceMetric(
                name="Excessive Force Unwrapping",
                category=Category.MEMORY,
                severity=Severity.MEDIUM,
                file=file_name,
                line=None,
                description=f"Found {force_unwrap_count} force unwraps",
                recommendation="Use optional binding or nil-coalescing",
                estimated_impact=Impact.MEDIUM
            ))
    
    def _get_line_number(self, position: int, content: str) -> int:
        """Get line number from character position"""
        return content[:position].count('\n') + 1
    
    def _generate_summary(self) -> Dict:
        """Generate analysis summary"""
        severity_counts = {
            'critical': sum(1 for m in self.metrics if m.severity == Severity.CRITICAL),
            'high': sum(1 for m in self.metrics if m.severity == Severity.HIGH),
            'medium': sum(1 for m in self.metrics if m.severity == Severity.MEDIUM),
            'low': sum(1 for m in self.metrics if m.severity == Severity.LOW),
            'info': sum(1 for m in self.metrics if m.severity == Severity.INFO)
        }
        
        category_counts = {}
        for metric in self.metrics:
            category = metric.category.value
            category_counts[category] = category_counts.get(category, 0) + 1
        
        # Calculate performance score (0-100)
        total_weight = (
            severity_counts['critical'] * 10 +
            severity_counts['high'] * 5 +
            severity_counts['medium'] * 2 +
            severity_counts['low']
        )
        max_weight = len(self.metrics) * 10 if self.metrics else 1
        performance_score = max(0, 100 - (total_weight / max_weight * 100))
        
        # Generate recommendations
        recommendations = []
        
        if severity_counts['critical'] > 0:
            recommendations.append(f"Fix {severity_counts['critical']} critical issues immediately")
        
        if category_counts.get(Category.MEMORY.value, 0) > 3:
            recommendations.append("Review memory management patterns")
        
        if category_counts.get(Category.CONCURRENCY.value, 0) > 3:
            recommendations.append("Audit concurrency patterns for thread safety")
        
        if category_counts.get(Category.UI.value, 0) > 2:
            recommendations.append("Optimize UI operations for responsiveness")
        
        if performance_score < 70:
            recommendations.append("Consider performance profiling with Instruments")
        
        return {
            'total_issues': len(self.metrics),
            'severity_counts': severity_counts,
            'category_counts': category_counts,
            'performance_score': round(performance_score, 2),
            'recommendations': recommendations,
            'top_issues': self._get_top_issues()
        }
    
    def _get_top_issues(self, limit: int = 5) -> List[Dict]:
        """Get top priority issues"""
        # Sort by severity (critical first) then by impact
        severity_order = {
            Severity.CRITICAL: 0,
            Severity.HIGH: 1,
            Severity.MEDIUM: 2,
            Severity.LOW: 3,
            Severity.INFO: 4
        }
        
        sorted_metrics = sorted(
            self.metrics,
            key=lambda m: (severity_order[m.severity], m.estimated_impact.value)
        )
        
        return [m.to_dict() for m in sorted_metrics[:limit]]
    
    def export_json(self, result: Dict, output_path: str):
        """Export results to JSON"""
        with open(output_path, 'w') as f:
            json.dump(result, f, indent=2)
        print(f"✅ JSON report exported to: {output_path}")
    
    def export_markdown(self, result: Dict, output_path: str):
        """Export results to Markdown"""
        md_content = f"""# Performance Analysis Report

**Generated:** {result['timestamp']}
**Project:** {result['project_path']}
**Performance Score:** {result['summary']['performance_score']}/100

## Summary

- **Total Issues:** {result['summary']['total_issues']}
- **Critical:** {result['summary']['severity_counts']['critical']}
- **High:** {result['summary']['severity_counts']['high']}
- **Medium:** {result['summary']['severity_counts']['medium']}
- **Low:** {result['summary']['severity_counts']['low']}

## Issues by Category

"""
        
        for category, count in sorted(result['summary']['category_counts'].items(), 
                                     key=lambda x: x[1], reverse=True):
            md_content += f"- **{category}:** {count} issues\n"
        
        md_content += "\n## Recommendations\n\n"
        for rec in result['summary']['recommendations']:
            md_content += f"- {rec}\n"
        
        md_content += "\n## Top Priority Issues\n\n"
        for issue in result['summary']['top_issues']:
            md_content += f"""### {issue['name']}

- **File:** `{issue['file']}`{f" (line {issue['line']})" if issue['line'] else ""}
- **Severity:** {issue['severity']}
- **Category:** {issue['category']}
- **Impact:** {issue['estimated_impact']}
- **Description:** {issue['description']}
- **Recommendation:** {issue['recommendation'] or 'N/A'}

---

"""
        
        with open(output_path, 'w') as f:
            f.write(md_content)
        print(f"✅ Markdown report exported to: {output_path}")


def main():
    """Main execution function"""
    parser = argparse.ArgumentParser(description='Performance Analyzer for Swift Projects')
    parser.add_argument('project_path', help='Path to Swift project')
    parser.add_argument('-o', '--output', help='Output directory', 
                       default=None)
    parser.add_argument('--json', action='store_true', 
                       help='Export JSON report')
    parser.add_argument('--markdown', action='store_true', 
                       help='Export Markdown report')
    
    args = parser.parse_args()
    
    # Set output directory
    output_dir = args.output or os.path.join(args.project_path, 'performance_analysis')
    os.makedirs(output_dir, exist_ok=True)
    
    # Run analysis
    analyzer = PerformanceAnalyzer(args.project_path)
    result = analyzer.analyze()
    
    # Generate timestamp for file names
    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
    
    # Export reports
    if args.json or not (args.json or args.markdown):
        json_path = os.path.join(output_dir, f'performance_{timestamp}.json')
        analyzer.export_json(result, json_path)
    
    if args.markdown or not (args.json or args.markdown):
        md_path = os.path.join(output_dir, f'performance_{timestamp}.md')
        analyzer.export_markdown(result, md_path)
    
    # Print summary
    print(f"""
📊 Performance Analysis Complete!
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Performance Score: {result['summary']['performance_score']}/100
Total Issues: {result['summary']['total_issues']}
Critical: {result['summary']['severity_counts']['critical']} | High: {result['summary']['severity_counts']['high']} | Medium: {result['summary']['severity_counts']['medium']} | Low: {result['summary']['severity_counts']['low']}

Top Recommendations:""")
    
    for i, rec in enumerate(result['summary']['recommendations'], 1):
        print(f"{i}. {rec}")
    
    print(f"\nReports saved to: {output_dir}")


if __name__ == '__main__':
    main()

File: Scripts/PerformanceAnalyzer.swift
#!/usr/bin/env swift

import Foundation

/// A modular and reusable performance analyzer for Swift projects
/// This tool can be integrated into CI/CD pipelines or run standalone
class PerformanceAnalyzer {
    
    // MARK: - Data Structures
    
    struct PerformanceMetric: Codable {
        let name: String
        let category: Category
        let severity: Severity
        let file: String
        let line: Int?
        let description: String
        let recommendation: String?
        let estimatedImpact: Impact
        
        enum Category: String, Codable {
            case memory = "Memory Management"
            case concurrency = "Concurrency"
            case algorithm = "Algorithm Complexity"
            case io = "I/O Operations"
            case ui = "UI Responsiveness"
            case network = "Network"
            case database = "Database"
        }
        
        enum Severity: String, Codable {
            case critical = "Critical"
            case high = "High"
            case medium = "Medium"
            case low = "Low"
            case info = "Info"
        }
        
        enum Impact: String, Codable {
            case high = "High Performance Impact"
            case medium = "Medium Performance Impact"
            case low = "Low Performance Impact"
            case negligible = "Negligible"
        }
    }
    
    struct AnalysisResult: Codable {
        let timestamp: Date
        let projectPath: String
        let metrics: [PerformanceMetric]
        let summary: Summary
        
        struct Summary: Codable {
            let totalIssues: Int
            let criticalCount: Int
            let highCount: Int
            let mediumCount: Int
            let lowCount: Int
            let categoryCounts: [String: Int]
            let performanceScore: Double
            let recommendations: [String]
        }
    }
    
    // MARK: - Properties
    
    private let projectPath: String
    private var metrics: [PerformanceMetric] = []
    private let fileManager = FileManager.default
    
    // MARK: - Patterns to Detect
    
    private struct Patterns {
        static let strongReferencePatterns = [
            "\\{\\s*\\[\\s*(?:weak|unowned)\\s+self\\s*\\]",  // Good pattern
            "\\{\\s*\\[self\\]",  // Potential retain cycle
            "self\\.",  // Direct self reference
            "timer\\.scheduledTimer",  // Timer retention
            "NotificationCenter\\.default\\.addObserver"  // Notification retention
        ]
        
        static let memoryPatterns = [
            "autoreleasepool",  // Memory pool usage
            "\\.copy\\(\\)",  // Copy operations
            "Data\\(contentsOf:",  // Large data loading
            "UIImage\\(named:",  // Image caching
            "NSCache",  // Cache usage
            "deinit\\s*\\{",  // Deinit implementation
            "memoryWarning"  // Memory warning handling
        ]
        
        static let concurrencyPatterns = [
            "DispatchQueue\\.main\\.async",  // Main thread dispatch
            "DispatchQueue\\.global",  // Background dispatch
            "Task\\.detached",  // Detached tasks
            "await\\s+withChecked",  // Checked continuations
            "actor\\s+",  // Actor usage
            "@MainActor",  // Main actor annotation
            "nonisolated",  // Nonisolated usage
            "Sendable"  // Sendable conformance
        ]
        
        static let algorithmPatterns = [
            "for.*in.*for.*in",  // Nested loops
            "\\.sorted\\(\\)",  // Sorting operations
            "\\.filter\\(.*\\.filter",  // Chained filters
            "\\.map\\(.*\\.map",  // Chained maps
            "\\.reduce\\(",  // Reduce operations
            "recursiv",  // Recursive calls
            "while.*true",  // Infinite loops
            "\\.contains\\(where:"  // Complex contains
        ]
        
        static let ioPatterns = [
            "FileManager\\.default",  // File operations
            "try.*Data\\(contentsOf",  // Synchronous file loading
            "UserDefaults\\.standard",  // UserDefaults access
            "CoreData",  // Core Data operations
            "\\.write\\(to:",  // File writing
            "JSONEncoder\\(\\)",  // JSON encoding
            "JSONDecoder\\(\\)"  // JSON decoding
        ]
    }
    
    // MARK: - Initialization
    
    init(projectPath: String) {
        self.projectPath = projectPath
    }
    
    // MARK: - Analysis Methods
    
    func analyze() -> AnalysisResult {
        print("🔍 Starting performance analysis for: \(projectPath)")
        
        // Clear previous metrics
        metrics.removeAll()
        
        // Analyze Swift files
        let swiftFiles = findSwiftFiles()
        print("📁 Found \(swiftFiles.count) Swift files to analyze")
        
        for file in swiftFiles {
            analyzeFile(at: file)
        }
        
        // Generate summary
        let summary = generateSummary()
        
        // Create result
        let result = AnalysisResult(
            timestamp: Date(),
            projectPath: projectPath,
            metrics: metrics,
            summary: summary
        )
        
        return result
    }
    
    private func findSwiftFiles() -> [String] {
        var swiftFiles: [String] = []
        
        if let enumerator = fileManager.enumerator(atPath: projectPath) {
            for case let file as String in enumerator {
                if file.hasSuffix(".swift") && 
                   !file.contains(".build") && 
                   !file.contains("Tests") {
                    swiftFiles.append("\(projectPath)/\(file)")
                }
            }
        }
        
        return swiftFiles
    }
    
    private func analyzeFile(at path: String) {
        guard let content = try? String(contentsOfFile: path, encoding: .utf8) else {
            return
        }
        
        let fileName = (path as NSString).lastPathComponent
        let lines = content.components(separatedBy: .newlines)
        
        // Analyze memory management
        analyzeMemoryManagement(content: content, fileName: fileName, lines: lines)
        
        // Analyze concurrency
        analyzeConcurrency(content: content, fileName: fileName, lines: lines)
        
        // Analyze algorithm complexity
        analyzeAlgorithmComplexity(content: content, fileName: fileName, lines: lines)
        
        // Analyze I/O operations
        analyzeIOOperations(content: content, fileName: fileName, lines: lines)
        
        // Analyze UI responsiveness
        analyzeUIResponsiveness(content: content, fileName: fileName, lines: lines)
    }
    
    private func analyzeMemoryManagement(content: String, fileName: String, lines: [String]) {
        // Check for potential retain cycles
        if content.contains("{ [self]") || content.contains("{[self]") {
            if let lineNumber = findLineNumber(pattern: "\\{\\s*\\[self\\]", in: lines) {
                metrics.append(PerformanceMetric(
                    name: "Potential Retain Cycle",
                    category: .memory,
                    severity: .high,
                    file: fileName,
                    line: lineNumber,
                    description: "Strong self capture in closure may cause retain cycle",
                    recommendation: "Use [weak self] or [unowned self] to break potential retain cycles",
                    estimatedImpact: .high
                ))
            }
        }
        
        // Check for large data loading
        if content.contains("Data(contentsOf:") && !content.contains("async") {
            if let lineNumber = findLineNumber(pattern: "Data\\(contentsOf:", in: lines) {
                metrics.append(PerformanceMetric(
                    name: "Synchronous Data Loading",
                    category: .memory,
                    severity: .medium,
                    file: fileName,
                    line: lineNumber,
                    description: "Synchronous data loading can block main thread and consume memory",
                    recommendation: "Use async/await or background queue for data loading",
                    estimatedImpact: .medium
                ))
            }
        }
        
        // Check for missing deinit in classes with resources
        if content.contains("class ") && 
           (content.contains("Timer") || content.contains("NotificationCenter")) &&
           !content.contains("deinit") {
            metrics.append(PerformanceMetric(
                name: "Missing Deinit",
                category: .memory,
                severity: .medium,
                file: fileName,
                line: nil,
                description: "Class with resources lacks deinit implementation",
                recommendation: "Implement deinit to properly clean up resources",
                estimatedImpact: .medium
            ))
        }
    }
    
    private func analyzeConcurrency(content: String, fileName: String, lines: [String]) {
        // Check for main thread blocking
        if content.contains("DispatchQueue.main.sync") {
            if let lineNumber = findLineNumber(pattern: "DispatchQueue\\.main\\.sync", in: lines) {
                metrics.append(PerformanceMetric(
                    name: "Main Thread Blocking",
                    category: .concurrency,
                    severity: .critical,
                    file: fileName,
                    line: lineNumber,
                    description: "Synchronous dispatch to main queue can cause deadlock",
                    recommendation: "Use async dispatch or restructure code to avoid blocking",
                    estimatedImpact: .high
                ))
            }
        }
        
        // Check for unstructured concurrency
        if content.contains("Task.detached") {
            if let lineNumber = findLineNumber(pattern: "Task\\.detached", in: lines) {
                metrics.append(PerformanceMetric(
                    name: "Unstructured Concurrency",
                    category: .concurrency,
                    severity: .low,
                    file: fileName,
                    line: lineNumber,
                    description: "Detached tasks lose structured concurrency benefits",
                    recommendation: "Consider using Task {} or async methods for structured concurrency",
                    estimatedImpact: .low
                ))
            }
        }
        
        // Check for missing @MainActor on UI updates
        if content.contains("@Published") && !content.contains("@MainActor") {
            metrics.append(PerformanceMetric(
                name: "UI State Without MainActor",
                category: .concurrency,
                severity: .medium,
                file: fileName,
                line: nil,
                description: "Published properties should be isolated to MainActor",
                recommendation: "Add @MainActor annotation to ensure UI updates on main thread",
                estimatedImpact: .medium
            ))
        }
    }
    
    private func analyzeAlgorithmComplexity(content: String, fileName: String, lines: [String]) {
        // Check for nested loops
        let nestedLoopPattern = "for.*\\{[^}]*for.*\\{"
        if let range = content.range(of: nestedLoopPattern, options: .regularExpression) {
            if let lineNumber = findLineNumber(pattern: nestedLoopPattern, in: lines) {
                metrics.append(PerformanceMetric(
                    name: "Nested Loop Detected",
                    category: .algorithm,
                    severity: .medium,
                    file: fileName,
                    line: lineNumber,
                    description: "Nested loops can lead to O(n²) complexity",
                    recommendation: "Consider optimizing with hash maps or better algorithms",
                    estimatedImpact: .medium
                ))
            }
        }
        
        // Check for multiple chained operations
        if content.contains(".filter") && content.contains(".map") && content.contains(".reduce") {
            metrics.append(PerformanceMetric(
                name: "Multiple Collection Operations",
                category: .algorithm,
                severity: .low,
                file: fileName,
                line: nil,
                description: "Multiple chained collection operations create intermediate arrays",
                recommendation: "Consider using lazy collections or combining operations",
                estimatedImpact: .low
            ))
        }
    }
    
    private func analyzeIOOperations(content: String, fileName: String, lines: [String]) {
        // Check for synchronous file operations
        if content.contains("try ") && content.contains("contentsOf") && !content.contains("async") {
            metrics.append(PerformanceMetric(
                name: "Synchronous File I/O",
                category: .io,
                severity: .medium,
                file: fileName,
                line: nil,
                description: "Synchronous file operations can block thread",
                recommendation: "Use async/await for file operations",
                estimatedImpact: .medium
            ))
        }
        
        // Check for excessive UserDefaults access
        let userDefaultsCount = content.components(separatedBy: "UserDefaults.standard").count - 1
        if userDefaultsCount > 5 {
            metrics.append(PerformanceMetric(
                name: "Excessive UserDefaults Access",
                category: .io,
                severity: .low,
                file: fileName,
                line: nil,
                description: "Multiple UserDefaults accesses detected (\(userDefaultsCount) times)",
                recommendation: "Cache UserDefaults values in properties",
                estimatedImpact: .low
            ))
        }
    }
    
    private func analyzeUIResponsiveness(content: String, fileName: String, lines: [String]) {
        // Check for heavy operations in view body
        if fileName.contains("View.swift") && content.contains("struct") && content.contains("View") {
            if content.contains(".sorted()") || content.contains(".filter") {
                metrics.append(PerformanceMetric(
                    name: "Heavy Operation in View",
                    category: .ui,
                    severity: .high,
                    file: fileName,
                    line: nil,
                    description: "Heavy operations in SwiftUI view body",
                    recommendation: "Move heavy operations to ViewModel or use @State for caching",
                    estimatedImpact: .high
                ))
            }
        }
        
        // Check for missing async image loading
        if content.contains("UIImage(data:") || content.contains("Image(uiImage:") {
            if !content.contains("async") && !content.contains("AsyncImage") {
                metrics.append(PerformanceMetric(
                    name: "Synchronous Image Loading",
                    category: .ui,
                    severity: .medium,
                    file: fileName,
                    line: nil,
                    description: "Synchronous image loading can freeze UI",
                    recommendation: "Use AsyncImage or async loading pattern",
                    estimatedImpact: .medium
                ))
            }
        }
    }
    
    private func findLineNumber(pattern: String, in lines: [String]) -> Int? {
        for (index, line) in lines.enumerated() {
            if line.range(of: pattern, options: .regularExpression) != nil {
                return index + 1
            }
        }
        return nil
    }
    
    private func generateSummary() -> AnalysisResult.Summary {
        let criticalCount = metrics.filter { $0.severity == .critical }.count
        let highCount = metrics.filter { $0.severity == .high }.count
        let mediumCount = metrics.filter { $0.severity == .medium }.count
        let lowCount = metrics.filter { $0.severity == .low }.count
        
        var categoryCounts: [String: Int] = [:]
        for metric in metrics {
            categoryCounts[metric.category.rawValue, default: 0] += 1
        }
        
        // Calculate performance score (0-100)
        let totalWeight = Double(criticalCount * 10 + highCount * 5 + mediumCount * 2 + lowCount)
        let maxWeight = Double(metrics.count * 10)
        let performanceScore = maxWeight > 0 ? max(0, 100 - (totalWeight / maxWeight * 100)) : 100
        
        // Generate top recommendations
        var recommendations: [String] = []
        
        if criticalCount > 0 {
            recommendations.append("Fix \(criticalCount) critical performance issues immediately")
        }
        
        if categoryCounts[PerformanceMetric.Category.memory.rawValue, default: 0] > 3 {
            recommendations.append("Review memory management patterns and potential leaks")
        }
        
        if categoryCounts[PerformanceMetric.Category.concurrency.rawValue, default: 0] > 3 {
            recommendations.append("Audit concurrency patterns for thread safety")
        }
        
        if categoryCounts[PerformanceMetric.Category.ui.rawValue, default: 0] > 2 {
            recommendations.append("Optimize UI operations for better responsiveness")
        }
        
        if performanceScore < 70 {
            recommendations.append("Consider performance profiling with Instruments")
        }
        
        return AnalysisResult.Summary(
            totalIssues: metrics.count,
            criticalCount: criticalCount,
            highCount: highCount,
            mediumCount: mediumCount,
            lowCount: lowCount,
            categoryCounts: categoryCounts,
            performanceScore: performanceScore.rounded(),
            recommendations: recommendations
        )
    }
    
    // MARK: - Export Methods
    
    func exportJSON(result: AnalysisResult, to path: String) throws {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        
        let data = try encoder.encode(result)
        try data.write(to: URL(fileURLWithPath: path))
        print("✅ JSON report exported to: \(path)")
    }
    
    func exportMarkdown(result: AnalysisResult, to path: String) throws {
        var markdown = """
        # Performance Analysis Report
        
        **Generated:** \(ISO8601DateFormatter().string(from: result.timestamp))
        **Project:** \(result.projectPath)
        **Performance Score:** \(result.summary.performanceScore)/100
        
        ## Summary
        
        - **Total Issues:** \(result.summary.totalIssues)
        - **Critical:** \(result.summary.criticalCount)
        - **High:** \(result.summary.highCount)
        - **Medium:** \(result.summary.mediumCount)
        - **Low:** \(result.summary.lowCount)
        
        ## Issues by Category
        
        """
        
        for (category, count) in result.summary.categoryCounts.sorted(by: { $0.value > $1.value }) {
            markdown += "- **\(category):** \(count) issues\n"
        }
        
        markdown += "\n## Recommendations\n\n"
        for recommendation in result.summary.recommendations {
            markdown += "- \(recommendation)\n"
        }
        
        markdown += "\n## Detailed Findings\n\n"
        
        // Group metrics by severity
        let groupedMetrics = Dictionary(grouping: result.metrics) { $0.severity }
        
        for severity in [PerformanceMetric.Severity.critical, .high, .medium, .low] {
            if let severityMetrics = groupedMetrics[severity], !severityMetrics.isEmpty {
                markdown += "### \(severity.rawValue) Priority\n\n"
                
                for metric in severityMetrics {
                    markdown += """
                    #### \(metric.name)
                    
                    - **File:** `\(metric.file)`\(metric.line.map { " (line \($0))" } ?? "")
                    - **Category:** \(metric.category.rawValue)
                    - **Impact:** \(metric.estimatedImpact.rawValue)
                    - **Description:** \(metric.description)
                    \(metric.recommendation.map { "- **Recommendation:** \($0)" } ?? "")
                    
                    ---
                    
                    """
                }
            }
        }
        
        try markdown.write(to: URL(fileURLWithPath: path), atomically: true, encoding: .utf8)
        print("✅ Markdown report exported to: \(path)")
    }
}

// MARK: - Main Execution

let arguments = CommandLine.arguments

if arguments.count < 2 {
    print("Usage: swift PerformanceAnalyzer.swift <project_path> [output_path]")
    exit(1)
}

let projectPath = arguments[1]
let outputPath = arguments.count > 2 ? arguments[2] : "\(projectPath)/performance_analysis"

// Create output directory
try? FileManager.default.createDirectory(atPath: outputPath, withIntermediateDirectories: true)

// Run analysis
let analyzer = PerformanceAnalyzer(projectPath: projectPath)
let result = analyzer.analyze()

// Export results
let timestamp = ISO8601DateFormatter().string(from: Date()).replacingOccurrences(of: ":", with: "-")
try? analyzer.exportJSON(result: result, to: "\(outputPath)/performance_\(timestamp).json")
try? analyzer.exportMarkdown(result: result, to: "\(outputPath)/performance_\(timestamp).md")

// Print summary
print("""

📊 Performance Analysis Complete!
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Performance Score: \(result.summary.performanceScore)/100
Total Issues: \(result.summary.totalIssues)
Critical: \(result.summary.criticalCount) | High: \(result.summary.highCount) | Medium: \(result.summary.mediumCount) | Low: \(result.summary.lowCount)

Top Recommendations:
""")

for (index, recommendation) in result.summary.recommendations.enumerated() {
    print("\(index + 1). \(recommendation)")
}

print("\nReports saved to: \(outputPath)")
